internal/controllers/topology/cluster/desired_state_test.go

```diff
@@ -1603,6 +1606,105 @@ func TestComputeMachineDeployment(t *testing.T) {
 		// Check that UnhealthyConditions are set as expected.
 		g.Expect(actual.MachineHealthCheck.Spec.UnhealthyConditions).To(Equal(unhealthyConditions))
 	})
+	t.Run("Should correctly set the finalizer", func(t *testing.T) {
+		g := NewWithT(t)
+		s := scope.New(cluster)
+		s.Blueprint = blueprint
+		deletionTimestamp := metav1.Now()
+		currentMd := &clusterv1.MachineDeployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "existing-deployment-1",
+			},
+			Spec: clusterv1.MachineDeploymentSpec{
+				Replicas: pointer.Int32(3),
+				Template: clusterv1.MachineTemplateSpec{
+					Spec: clusterv1.MachineSpec{
+						Version: pointer.String("v1.21.2"),
+						Bootstrap: clusterv1.Bootstrap{
+							ConfigRef: contract.ObjToRef(workerBootstrapTemplate),
+						},
+						InfrastructureRef: *contract.ObjToRef(workerInfrastructureMachineTemplate),
+					},
+				},
+			},
+		}
+		mdTopology = clusterv1.MachineDeploymentTopology{
+			Metadata: clusterv1.ObjectMeta{
+				Labels: map[string]string{"foo": "baz"},
+			},
+			Class: "linux-worker",
+			Name:  "big-pool-of-machines",
+		}
+		t.Run("Should set the finalizer on a new MD", func(t *testing.T) {
+			s.Current.MachineDeployments = map[string]*scope.MachineDeploymentState{
+				"big-pool-of-machines": {
+					Object:                        currentMd,
+					BootstrapTemplate:             workerBootstrapTemplate,
+					InfrastructureMachineTemplate: workerInfrastructureMachineTemplate,
+				},
+			}
+			actual, err := computeMachineDeployment(ctx, s, nil, mdTopology)
+			g.Expect(err).ToNot(HaveOccurred())
+			actualMd := actual.Object
+			g.Expect(controllerutil.ContainsFinalizer(actualMd, clusterv1.MachineDeploymentTopologyFinalizer)).To(BeTrue())
+		},
+		)
+		t.Run("Should retain finalizer on an existing MD if not set", func(t *testing.T) {
+			s.Current.MachineDeployments = map[string]*scope.MachineDeploymentState{}
+			actual, err := computeMachineDeployment(ctx, s, nil, mdTopology)
+			g.Expect(err).ToNot(HaveOccurred())
+			actualMd := actual.Object
+			g.Expect(controllerutil.ContainsFinalizer(actualMd, clusterv1.MachineDeploymentTopologyFinalizer)).To(BeTrue())
+		},
+		)
+		t.Run("Should preserve finalizer on an existing MD if set", func(t *testing.T) {
+			s.Current.MachineDeployments = map[string]*scope.MachineDeploymentState{
+				"big-pool-of-machines": {
+					Object:                        currentMd,
+					BootstrapTemplate:             workerBootstrapTemplate,
+					InfrastructureMachineTemplate: workerInfrastructureMachineTemplate,
+				},
+			}
+			currentMd.Finalizers = []string{clusterv1.MachineDeploymentTopologyFinalizer}
```

Comment:
Do we need deepcopy to 100% avoid mutations to an object shared across test cases?

Reply:
I know it works as we are probably setting fields in the right order across test cases but I would really prefer not to depend on this

Reply:
Good idea!